File: problems-adapt/0139_vocabulary-segmentation/solution_dp.java
import java.util.HashSet;
import java.util.Set;

class Solution {

    public boolean canSegment(String s, String[] vocabulary) {
        Set<String> words = new HashSet<>();
        for (String word : vocabulary) {
            words.add(word);
        }
        int n = s.length();
        // reachable[i]: the prefix s[0..i) can be segmented; the empty prefix
        // is trivially segmentable.
        boolean[] reachable = new boolean[n + 1];
        reachable[0] = true;
        for (int i = 1; i <= n; i++) {
            // Any segmentation of s[0..i) ends with a last word s[j..i).
            for (int j = 0; j < i; j++) {
                if (reachable[j] && words.contains(s.substring(j, i))) {
                    reachable[i] = true;
                    // Only feasibility matters, so stop at the first split.
                    break;
                }
            }
        }
        return reachable[n];
    }
}

File: problems-adapt/0000-0100/0052_vocabulary-segmentation/solution_dp.go
func canSegment(s string, vocabulary []string) bool {
	words := make(map[string]struct{}, len(vocabulary))
	for _, word := range vocabulary {
		words[word] = struct{}{}
	}
	n := len(s)
	// reachable[i]: the prefix s[0..i) can be segmented; the empty prefix is
	// trivially segmentable.
	reachable := make([]bool, n+1)
	reachable[0] = true
	for i := 1; i <= n; i++ {
		// Any segmentation of s[0..i) ends with a last word s[j..i).
		for j := 0; j < i; j++ {
			if reachable[j] {
				if _, ok := words[s[j:i]]; ok {
					reachable[i] = true
					// Only feasibility matters, so stop at the first split.
					break
				}
			}
		}
	}
	return reachable[n]
}
